[cutlass backend] update try_import_cutlass to accomodate for pip install

henrylhtsang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #145891

The goal of this PR is to provide 3 ways for people to try out CUTLASS backend:
1. fbcode / internal 
2. pip install torch (nightly) and pip install nvidia-cutlass 
3. build from source

I will go into more detailed combos between building from source and downloading via pip for torch and cutlass. 

repro:
```
import torch 
import torch.nn as nn

import torch._inductor.config as config 

config.force_disable_caches = True
config.max_autotune = True
config.max_autotune_gemm_backends = "CUTLASS"
# the following is only needed if you use a custom cutlass library 
# config.cuda.cutlass_dir = "/data/users/henrylhtsang/cutlass"

class TestModule(nn.Module):
    def forward(self, A, B):
        return A @ B 

model = TestModule().cuda()
M, K, N = 2048, 2048, 2048
A = torch.randn(M, K).cuda().half()
B = torch.randn(K, N).cuda().half()

C = torch.compile(model, fullgraph=True)(A, B)
```

## pre-requisite 
Assuming you have the right cuda toolkit. Recommend 12.4. Make sure PATH, LD_LIBRARY_PATH and CUDA_NVCC_EXECUTABLE are good. 

## combo 1: pip install torch + pip install nvidia-cutlass
Check https://pytorch.org/get-started/locally/ for **nightly** install command. 
```
pip install --pre torch --index-url https://download.pytorch.org/whl/nightly/cu124
pip install nvidia-cutlass
```
Then try running the script above. It should work. 

## combo 2: build torch from source + pip install nvidia-cutlass
This is going to be be pretty straightforward. Just keep in mind that even though pytorch/third_party/cutlass exists, the one that will be used is the pip package, so mindful of version differences.

## combo 3: build torch from source + use pytorch/third_party/cutlass
This is how most pytorch devs would do it. Just make sure you don't have a cutlass pip package installed, i.e., make sure `import cutlass_library` would fail on its own. 

## combo 4: any torch version + cutlass library from somewhere else
This is probably the only case you need to pass in cutlass_dir. Just set cutlass_dir to the cutlass repo library. The expectations is that cutlass_dir is the directory that contains include, tool, and python/cutlass_library.



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

henrylhtsang committed:
[cutlass backend] update try_import_cutlass to accomodate for pip install

[ghstack-poisoned]

henrylhtsang committed:
Update on "[cutlass backend] update try_import_cutlass to accomodate for pip install"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

Chillee commented:
I think this makes sense. It'd be good to have this path tested in CI tbh.

drisspg commented on torch/_inductor/codegen/cuda/cutlass_utils.py:
```diff
@@ -32,9 +32,39 @@ def _rename_cutlass_import(content: str, cutlass_modules: list[str]) -> str:
         return True
 
+    try:
+        import cutlass_library  # noqa: F401
```
Should we do some version checks?

henrylhtsang replied:
> Should we do some version checks?

@drisspg personally I think no version check is fine, since user might want to use a newer or older cutlass library? But see also my comment below

henrylhtsang replied:
Added a warning to check its at least 3.7.

henrylhtsang commented:
Question for @Chillee and @drisspg, do we expect cutlass to become a (nvidia-specific) dependency, or do we have to keep doing third-party?

Chillee commented:
@henrylhtsang well, there's a couple questions here about dependencies. For example, would we eventually want to install nvidia-cutlass by default with the pytorch package? I think the main question mark there is about nvcc as a dependency.

For now I think we should just allow folks to optionally install `nvidia-cutlass` from pip + their system install of nvcc to make it work.

henrylhtsang committed:
Update on "[cutlass backend] update try_import_cutlass to accomodate for pip install"


The goal of this PR is to provide 3 ways for people to try out CUTLASS backend:
1. fbcode / internal 
2. pip install torch (nightly) and pip install nvidia-cutlass 
3. build from source

I will go into more detailed combos between building from source and downloading via pip for torch and cutlass. 

repro:
```
import torch 
import torch.nn as nn

import torch._inductor.config as config 

config.force_disable_caches = True
config.max_autotune = True
config.max_autotune_gemm_backends = "CUTLASS"
# the following is only needed if you use a custom cutlass library 
# config.cuda.cutlass_dir = "/data/users/henrylhtsang/cutlass"

class TestModule(nn.Module):
    def forward(self, A, B):
        return A @ B 

model = TestModule().cuda()
M, K, N = 2048, 2048, 2048
A = torch.randn(M, K).cuda().half()
B = torch.randn(K, N).cuda().half()

C = torch.compile(model, fullgraph=True)(A, B)
```

## pre-requisite 
Assuming you have the right cuda toolkit. Recommend 12.4. Make sure PATH, LD_LIBRARY_PATH and CUDA_NVCC_EXECUTABLE are good. 

## combo 1: pip install torch + pip install nvidia-cutlass
Check https://pytorch.org/get-started/locally/ for **nightly** install command. 
```
pip install --pre torch --index-url https://download.pytorch.org/whl/nightly/cu124
pip install nvidia-cutlass
```
Then try running the script above. It should work. 

## combo 2: build torch from source + pip install nvidia-cutlass
This is going to be be pretty straightforward. Just keep in mind that even though pytorch/third_party/cutlass exists, the one that will be used is the pip package, so mindful of version differences.

## combo 3: build torch from source + use pytorch/third_party/cutlass
This is how most pytorch devs would do it. Just make sure you don't have a cutlass pip package installed, i.e., make sure `import cutlass_library` would fail on its own. 

## combo 4: any torch version + cutlass library from somewhere else
This is probably the only case you need to pass in cutlass_dir. Just set cutlass_dir to the cutlass repo library. The expectations is that cutlass_dir is the directory that contains include, tool, and python/cutlass_library.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov 

[ghstack-poisoned]

henrylhtsang committed:
Update on "[cutlass backend] update try_import_cutlass to accomodate for pip install"


The goal of this PR is to provide 3 ways for people to try out CUTLASS backend:
1. fbcode / internal 
2. pip install torch (nightly) and pip install nvidia-cutlass 
3. build from source

I will go into more detailed combos between building from source and downloading via pip for torch and cutlass. 

repro:
```
import torch 
import torch.nn as nn

import torch._inductor.config as config 

config.force_disable_caches = True
config.max_autotune = True
config.max_autotune_gemm_backends = "CUTLASS"
# the following is only needed if you use a custom cutlass library 
# config.cuda.cutlass_dir = "/data/users/henrylhtsang/cutlass"

class TestModule(nn.Module):
    def forward(self, A, B):
        return A @ B 

model = TestModule().cuda()
M, K, N = 2048, 2048, 2048
A = torch.randn(M, K).cuda().half()
B = torch.randn(K, N).cuda().half()

C = torch.compile(model, fullgraph=True)(A, B)
```

## pre-requisite 
Assuming you have the right cuda toolkit. Recommend 12.4. Make sure PATH, LD_LIBRARY_PATH and CUDA_NVCC_EXECUTABLE are good. 

## combo 1: pip install torch + pip install nvidia-cutlass
Check https://pytorch.org/get-started/locally/ for **nightly** install command. 
```
pip install --pre torch --index-url https://download.pytorch.org/whl/nightly/cu124
pip install nvidia-cutlass
```
Then try running the script above. It should work. 

## combo 2: build torch from source + pip install nvidia-cutlass
This is going to be be pretty straightforward. Just keep in mind that even though pytorch/third_party/cutlass exists, the one that will be used is the pip package, so mindful of version differences.

## combo 3: build torch from source + use pytorch/third_party/cutlass
This is how most pytorch devs would do it. Just make sure you don't have a cutlass pip package installed, i.e., make sure `import cutlass_library` would fail on its own. 

## combo 4: any torch version + cutlass library from somewhere else
This is probably the only case you need to pass in cutlass_dir. Just set cutlass_dir to the cutlass repo library. The expectations is that cutlass_dir is the directory that contains include, tool, and python/cutlass_library.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov 

[ghstack-poisoned]

henrylhtsang commented:
@pytorchbot merge
